elasticdl/python/common/model_handler.py

```diff
@@ -0,0 +1,111 @@
+import abc
+
+import tensorflow as tf
+
+from elasticdl.python.common.log_utils import default_logger as logger
+from elasticdl.python.elasticdl.layers.embedding import Embedding
+
+
+class ModelHandler(metaclass=abc.ABCMeta):
+    """
+    Generate the model to train in ElasticDL for different distributed
+    strategies and export trained model in ElasticDL to SavedModel.
+    """
+
+    @abc.abstractmethod
+    def generate_train_model_for_elasticdl(self, model):
+        """
+        Generate a model to train in ElasticDL.
+        """
+        pass
+
+    @abc.abstractmethod
+    def get_saved_model_from_trained_model(self, model, dataset):
+        """
+        Get the model which can be exported a SavedModel
+        by tf.saved_model.save.
+        """
+        pass
+
+    @classmethod
+    def get_model_handler(cls, distribution_strategy=None):
+        """
+        Create a model handler to process the model for the
+        distributed strategy.
+        """
+        if distribution_strategy == "ParameterServerStrategy":
+            return ParameterServerModelHandler()
+        else:
+            return DefaultModelHandler()
+
+
+class DefaultModelHandler(ModelHandler):
+    """
+    Return the origin model to train and export.
+    """
+
+    def generate_train_model_for_elasticdl(self, model):
+        return model
+
+    def get_saved_model_from_trained_model(self, model, dataset):
+        if not model.inputs:
+            model._build_model_with_inputs(inputs=dataset, targets=None)
+        return model
+
+
+class ParameterServerModelHandler(ModelHandler):
+    def generate_train_model_for_elasticdl(self, model):
+        """
+        Replace the tf.keras.layers.Embedding layer in the model with
+        an elasticdl.layers.Embedding layer in ParameterServerStrategy.
+        """
+        if type(model) == tf.keras.Sequential or model._is_graph_network:
+            model = self._clone_model_for_sequential_and_functional(model)
+        else:
+            model = self._replace_embedding_attribute_for_subclass(model)
+        return model
+
+    def get_saved_model_from_trained_model(self, model, dataset):
+        """
+        To export model for tf-serving by tf.saved_model.save:
+        1. Add inputs and outputs to the model.
+        2. Restore Keras model and replace embedding parameters with trained
+        model.
+        """
+        # TODO
+        pass
+
+    def _clone_model_for_sequential_and_functional(self, model):
+        """
+        Clone a new model with elasticdl.layers.Embedding for
+        Sequential and functional API model.
+        """
+
+        def _clone_function(layer):
+            if type(layer) == tf.keras.layers.Embedding:
+                logger.info(
+                    "Replace Keras Embedding with \
+                            ElasticDL Embedding"
+                )
+                edl_embedding_layer = Embedding(
+                    output_dim=layer.output_dim,
```

Comment:
We need add input_dim in edl_embedding_layer. If user define the model using native keras embedding layer, user expect the vocab size = input_dim. edl_embedding_layer should return error if the id > input_dim.